Quality-of-life API improvements.

arnsholt opened the pull request:
As we discussed in #31, From<> implementations for the types that seem to make sense (for EliasFano `.iter().collect()` should suffice, and the RMQ types are Deref and aren't constructed from BitVecs). While using the library I've also felt the need for the extend and split functionality, so I added those while I was at it.

Adds:
- `From<T> for BitVec` implementations for converting RsVec, BpTree and WaveletMatrix back into BitVecs
- `impl Extend<BitVec> for BitVec` and `BitVec.extend_vec()` to append many bits at once
- `BitVec.split_at()` and `._split_at_unchecked()` to split a BitVec into two halves

Cydhra committed:
added `into_bit_vec` to RsVec (#31)

Cydhra committed:
add example to into_bit_vec

Cydhra committed:
add `into_parentheses_vec` to `BpTree` (#31)

Cydhra committed:
add `must_use` to `into_parentheses_vec`

Cydhra committed:
reduced branching in get_bits, improving performance of various data structures

Cydhra committed:
same branching reduction in BitVec::get_bits_unchecked

arnsholt committed:
Quality-of-life API improvements.

Adds:
- `From<T> for BitVec` implementations for converting RsVec, BpTree and
  WaveletMatrix back into BitVecs
- `impl Extend<BitVec> for BitVec` and `BitVec.extend_vec()` to append
  many bits at once
- `BitVec.split_at()` and `._split_at_unchecked()` to split a BitVec
  into two halves

arnsholt commented:
And I see now that between me looking at this again yesterday and deciding to do it, and actually doing it today, you did most of it anyways. Oops. =)

Cydhra commented:
Well it doesn't matter, too much, the main work I did yesterday was fix the padding thing in b478799e3a26a6a33c815fc5210316bd1e090831, which enables the `From` implementation to work in the first place. The work on the `issue_31` branch isn't very substantial. You could rebase onto the branch though, if you want, as I already did the documentation work for the conversion methods, so you could rewire your `From` implementations to call those methods and profit from the existing documentation. If you think this is redundant, we can also ignore the branch.

Cydhra commented on src/wavelet/mod.rs:
```diff
@@ -2053,6 +2053,14 @@ impl WaveletMatrix {
     }
 }
 
+impl From<WaveletMatrix> for BitVec {
```
I don't understand the significance of this. A wavelet matrix encodes symbols bit per bit, the n'th bit of each symbol in the n'th RsVec of the WaveletMatrix. Why would we want to append all those vectors into a single vector?

arnsholt replied:
It's not something I've needed in my project, I added mostly for symmetry with BpTree. Since both are constructed from a BitVec, I figured it made sense to implement the backwards conversion for both types. If that's nonsense for WaveletMatrix we can delete it.

Cydhra replied:
Yeah it doesn't make sense here. The wavelet matrix is constructed from integers, or for very strange inputs from BitVecs, but they are transposed during construction, so appending them like this doesn't make sense.

Cydhra commented on src/bit_vec/mod.rs:
```diff
@@ -736,6 +736,19 @@ impl BitVec {
     }
 
+    /// Append the bits of another bit vector to the vector.
+    pub fn extend_bitvec(&mut self, other: &Self) {
```
We need test cases for all the methods. I can implement them if you enabled the option that I can edit the PR.

arnsholt committed:
Change From implementations to use methods on their types.

arnsholt commented:
I rebased onto your branch and updated the From impls for RsVec and BpTree. Left WaveletMatrix as is for now.

I'm pretty sure I added edit for you on the PR, so if you have a clear idea of what tests you want you can put them in. Otherwise, I can try to add some tests tomorrow.

Cydhra committed:
revert From<WaveletMatrix> implementation for BitVec

Cydhra committed:
implement From<BpTree> for RsVec

Cydhra committed:
specify returned bit order of `BitVec::get_bits`

Cydhra committed:
add test case for `BitVec::extend`

Cydhra committed:
ensure extend only reallocates once, update its documentation

Cydhra committed:
add error section to `split_at` documentation

Cydhra committed:
fix: split_at_unchecked doesn't drop second part from self

Cydhra committed:
fix: split_at_unchecked copies wrong partial limb at the start of the second part

Cydhra committed:
fix: if we split in the last limb, or in a single-limb vector, we could be copying past the vector, or copying the same data twice

Cydhra committed:
add test case for split_at

Cydhra committed:
add test for the checked version of split_at

Cydhra reviewed (approved):
I have added test cases and fixed three bugs in split_at. I also removed the `From<WaveletMatrix>` implementation.

Cydhra commented on src/bit_vec/mod.rs:
```diff
@@ -1194,6 +1215,71 @@ impl BitVec {
+
+        let iter_limbs = if leading_partial > 0 {
+            other.append_bits_unchecked(
+                self.data[first_limb] >> (WORD_SIZE - leading_partial),
```
This fails if the limb isn't `WORD_SIZE` long, see red test case.

arnsholt committed:
Fix split_at implementation, and add more tests for splitting.

arnsholt commented:
Wow, I clearly didn't think the splitting logic through properly. Sorry for rushing making the PR and making you find the bugs first! I had a proper think about the algorithm and ended up with a factoring where the initial logic was more or less sound, but with an added edge case that needed handling where the split point is inside a partial last limb of the input vector. I also added a few more test cases.

arnsholt commented:
And I completely failed to see the unchecked set of tests just above. Sorry for being so sloppy again!

arnsholt committed:
Early return in split if second half is empty.

arnsholt commented:
There, now it passes the full test suite here.

Cydhra committed:
cleanup tests, add test for empty vector

Cydhra committed:
slight doc adjustment
